Add Maps to OED

ak476519 opened the pull request:
Introducing maps that graph energy usage by meters and groups of meters. This is documented as an issue and fixes #469.

Inside the admin panel, an image can be uploaded. Then add corresponding GIS coordinates for at least 3 points to calibrate. After a map is created, its info can be modified from a maps tab in the admin panel.
 
The admin can also add GIS coordinates for available meters in the meters panel. The meter will be displayed on a map, if it's inside the map's boundaries. 

This fixes #469. The changes by merging development also fixes #527.

Note merging this changes the node modules. Note merging this changes the database configuration.

ak476519 committed:
Merge remote-tracking branch 'upstream/development' into mapChart

ak476519 committed:
design & set up function hierarchy for mapChart

ak476519 committed:
migrate map initiate functions,
define state-changing actions

ak476519 committed:
Add action types for mapInitiateContainer

ak476519 committed:
comment out features belong to non-goals

ak476519 committed:
Merge remote-tracking branch 'upstream/development' into mapChart

ak476519 committed:
resolve conflicts

ak476519 committed:
finish mapInitiation

ak476519 committed:
fix lint errors

ak476519 committed:
revert map source upload format back to .txt files

ak476519 committed:
add MPL headers to files

ak476519 committed:
fix upload issue, add connection to PlotlyChart

ak476519 committed:
update comments

ak476519 committed:
fix page layout

ak476519 committed:
refactored map states into one to resolve overhead
in passing image data URL around

ak476519 committed:
start working on generalizing calibration

ak476519 committed:
Migration complete

ak476519 committed:
added code to get gps coordinate input

ak476519 committed:
normalize calibration to fit any image dimension

ak476519 committed:
refactor code to use Redux for state management

ak476519 committed:
debugged dashboard, initiation and chart layout

ak476519 committed:
connect calibration calculation to redux

ak476519 committed:
add code to calculate max error of map scale

ak476519 committed:
fix bug in calibration result display

ak476519 committed:
disable map hover info and axis display

ak476519 committed:
update version of react-plotlyjs-ts to fix issue
in mouse event listener overflow

ak476519 committed:
changed calibration result templates so that numbers make more sense

ak476519 committed:
fix  bugs

ak476519 committed:
add database functions to store maps

ak476519 committed:
add update db function for maps

ak476519 committed:
incorporate spinner into maps,
TODO: rewrite this part in next debug session

ak476519 committed:
create middleware functions to connect to database

ak476519 committed:
add code to test communication with database

ak476519 committed:
added patches to convert data to point type in DB,
added test for patches

ak476519 committed:
fix point type patches

ak476519 committed:
changed variable names to make them more intuitive

ak476519 committed:
fix spinner, debug mapApi,
verify map data are correctly fetched from database

ak476519 committed:
refactor incorporate calibrated points in state

ak476519 committed:
refactor map state to accept more data from db

ak476519 committed:
move map calibration to header

ak476519 committed:
add new column 'displayable' to maps table

ak476519 committed:
add code to calculate mapScale from endpoints

ak476519 committed:
fix package error in package lock

ak476519 committed:
put points on the map from information saved to db
, debugging in progress

ak476519 committed:
removed redundant 'cancel' button

ak476519 committed:
fix bug in calibration functions

ak476519 committed:
debug and finalize the concept model

ak476519 committed:
make map state analogous to meter state

ak476519 committed:
debug function: calculating scales from endpoints

ak476519 committed:
make header present on admin's map panel

ak476519 committed:
small code improvements

ak476519 committed:
avoid detecting clicks outside of map,
refactor variables, add comments

ak476519 committed:
make mapState analogous of meterState

ak476519 committed:
fix bugs

ak476519 committed:
implement map details tab and mapChart selection

ak476519 committed:
resolve routing issue of entering calibration

ak476519 committed:
fix bugs in router and state management

ak476519 committed:
fix issues in creating a new map on the interface

ak476519 committed:
redirect user to home page after new map is saved

ak476519 committed:
fix bug in not seeing filename in map details

ak476519 committed:
fix bugs in submitting after calibration

ak476519 committed:
redirect user back to map panel after calibration

ak476519 committed:
remove unused props

ak476519 committed:
add cancel calibration and input validation,
put error on log when gps input validation fails

ak476519 committed:
fix bug in logs from client-side

ak476519 committed:
added unit test for mapsApi

ak476519 committed:
refined map routers

ak476519 committed:
remove displayable filter
to make map chart display analogous to meters

ak476519 committed:
modify meters table to store GPS coordinates,
modify meters panel to accept manual GPS coordinates input

ak476519 committed:
add comments to GPS points

ak476519 committed:
added gps point to Meters model functions,
tests all passed with changes

ak476519 committed:
disable selection of undisplayable meters
by using filters in drop-down menu,
add type checking in graphing maps;

ak476519 committed:
change to saving HTML image in Redux to fix bug:
when src of a new image is set by copying source URL from redux,
sometimes width and height will not be updated immediately.

ak476519 committed:
fix bug in front-end logging, and in delete button

ak476519 committed:
fix router's bug in rejecting a meter w/t gps,
fix lint errors

ak476519 committed:
Merge remote-tracking branch
'upstream/development' into mapChart

ak476519 committed:
add sql to alter table in migrations folder

ak476519 committed:
filter out maps without calibration data at home

ak476519 committed:
modify table format of map panel

ak476519 committed:
ask for confirmation before deleting a map

ak476519 committed:
fix build error

ak476519 committed:
fix lint errors

ak476519 committed:
add test for barchart compressed reading route

ak476519 committed:
set data fetched for maps to be
energy usage averaged by days

ak476519 committed:
fix bug in GPS display of meters panel

ak476519 committed:
set GPS point data to default as null

ak476519 committed:
add feature to display grids in calibration

ak476519 committed:
add support for redirecting to map by link

ak476519 committed:
make focus shift automatically to textarea
when editing

ak476519 commented:
**Note: further discussion is needed to decide what kind of data will be displayed on the map. 

Currently maps are set to display the same time interval selected in bars and lines, averaged by days. If the range is unbounded, it displays the average usage for the last 4 weeks'(28 days).

ak476519 commented:
Other features and changes: 

1. A logging API is added to document calibration process in the log, which includes actions and data input. 
2. Replaced the chart selection buttons with a dropdown menu to hold more options.

huss commented on src/server/migrations/0.5.0-0.6.0/preferences/sql/map/create_maps_table.sql:
```diff
@@ -0,0 +1,15 @@
+	note VARCHAR(500),
+	filename VARCHAR(200) NOT NULL,
+	modified_date TIMESTAMP NOT NULL,
+	origin POINT,
```
origin and opposite are DEFAULT NULL in create but not here.

huss commented on src/server/migrations/0.5.0-0.6.0/preferences/sql/map/create_maps_table.sql:
```diff
@@ -0,0 +1,15 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
```
For reasons I don't know, the directory structure is off for migrations. It should be src/server/migrations/0.50-0.6.0/sql (not src/server/migrations/0.50-0.6.0/preferences/sql). I don't think you created this but can you please fix this up.

huss commented on src/server/sql/meter/create_meters_table.sql:
```diff
@@ -3,10 +3,11 @@
-	meter_type meter_type NOT NULL
+	displayable BOOLEAN NOT NULL,
+	meter_type meter_type NOT NULL,
+	gps POINT
```
In migrations, gps is DEFAULT NULL but not here

huss commented on src/client/app/actions/logs.ts:
```diff
@@ -0,0 +1,33 @@
+import { logsApi } from "../utils/api";
+const moment = require('moment');
+
+export function logToServer(level: string, message: string, error?: Error, skipMail?: boolean) {
```
It would be nice to have some comment on what this does/is for.

huss commented on src/client/app/actions/map.ts:
```diff
@@ -0,0 +1,295 @@
+import * as moment from 'moment';
+import {browserHistory} from '../utils/history';
+import {logToServer} from "./logs";
+
```
Some comments would help.

huss replied:
good enough

huss commented on src/client/app/components/DashboardComponent.tsx:
```diff
@@ -7,18 +7,16 @@ import UIOptionsContainer from '../containers/UIOptionsContainer';
-import { Dispatch, Thunk, ActionType } from '../types/redux/actions';
 
-import * as Plotly from 'plotly.js';
 // TODO lowercase plotly for the import in index.d.ts but uppercase here.
```
I think this comment is now obsolete so can be removed?

huss commented on src/client/app/components/DashboardComponent.tsx:
```diff
@@ -56,11 +55,16 @@ export default class DashboardComponent extends React.Component<DashboardProps,
 				showSpinner = true;
 			}
 			ChartToRender = MultiCompareChartContainer;
+		} else {
```
This is similar logic to before but it does mean if someone makes a mistake in the chartToRender you get a map. At some point we should have a default that deals with errors?

ak476519 replied:
Changed to throwing an error if someone makes a mistake.

huss commented on src/client/app/components/maps/MapCalibrationComponent.tsx:
```diff
@@ -0,0 +1,53 @@
+		if (this.props.mode === CalibrationModeTypes.initiate) {
+			return (
+				<div className='container-fluid'>
+					<HeaderContainer />
```
Some /> have space  before and some do not. Reason?

huss commented on src/client/app/components/maps/MapCalibration_InfoDisplayComponent.tsx:
```diff
@@ -0,0 +1,111 @@
+			this.props.updateGPSCoordinates(gps);
+			this.resetInputField();
+		} else {
+			this.props.log('info', 'refused data point with invalid input');
```
Would be good if told invalid GPS provided in log msg.

huss commented on src/client/app/components/maps/MapCalibration_InitiateComponent.tsx:
```diff
@@ -0,0 +1,108 @@
+			}
+			await this.props.onSourceChange(source);
+		} catch (err) {
+			console.log(err);
```
Should this msg also include the context in which the error occurred.

huss commented on src/client/app/components/maps/MapViewComponent.tsx:
```diff
@@ -0,0 +1,301 @@
+	}
+
+	private formatDeleteButton() {
+
```
extra blank line.

huss commented on src/client/app/components/maps/MapViewComponent.tsx:
```diff
@@ -0,0 +1,301 @@
+		this.props.setCalibration(mode, this.props.id);
+	}
+}
+
```
extra blank lines.

huss commented on src/client/app/actions/mapReadings.ts:
```diff
@@ -0,0 +1,76 @@
+		const promises: Array<Promise<any>> = [];
+		const mapDuration = (timeInterval.equals(TimeInterval.unbounded()))? moment.duration(4, 'weeks')
+			: moment.duration(timeInterval.duration('days'), 'days');
+		console.log(mapDuration);
```
Is this for debugging (so should be removed) or should it be changed to use the logging system?

huss commented on src/client/app/containers/MapChartContainer.ts:
```diff
@@ -0,0 +1,145 @@
+			 */
+			const points: any[] = [];
+			for (const meterID of state.graph.selectedMeters) {
+				// console.log(state.meters.selectedMeters);
```
remove log statements?

huss commented on src/client/app/containers/MapChartContainer.ts:
```diff
@@ -0,0 +1,145 @@
+				width: image.width,
+				height: image.height
+			});
+			// map coordinates to individual traces, todo: finalize mapping function
```
Is this todo an issue?

huss commented on src/client/app/containers/MapChartContainer.ts:
```diff
@@ -0,0 +1,145 @@
+			x = points.map(point => (point.longitude - origin.longitude) / mapScale.degreePerUnitX);
+			y = points.map(point => (point.latitude - origin.latitude) / mapScale.degreePerUnitY);
+
+			const trace1 = {
```
I prefer not to have variable names that end with a number, esp. 1 since it looks like l.

huss replied:
Changed trace1 to traceOne

huss commented on src/client/app/containers/RouteContainer.ts:
```diff
@@ -12,7 +12,7 @@ import { changeOptionsFromLink, LinkOptions } from '../actions/graph';
 	return {
 		barStacking: state.graph.barStacking,
-		defaultLanguage: state.admin.defaultLanguage
+		defaultLanguage: state.admin.defaultLanguage,
```
I'm not generally commenting when lint shows an issue but this added comma is not needed.

huss commented on src/client/app/containers/maps/MapCalibration_ChartDisplayContainer.ts:
```diff
@@ -0,0 +1,165 @@
+	 *  for now, all of the points from dataPointTrace are ignored;
+ 	 */
+	const eligiblePoints = [];
+	for (const point of event.points) {
```
I'm not clear on why multiple points are in the event and then it seems you use the first one [0] in the end. Could you put in comment to explain this?

huss replied:
comment added by @ak476519 so done.

huss commented on src/client/app/containers/maps/MapCalibration_InfoDisplayContainer.ts:
```diff
@@ -0,0 +1,36 @@
+	const resultDisplay = (map.calibrationResult)?
+		`x: ${map.calibrationResult.maxError.x}%, y: ${map.calibrationResult.maxError.y}%`
+		: "Need more points";
+	const currentCartesianDisplay =(map.currentPoint)? `x: ${map.currentPoint.cartesian.x}, y: ${map.currentPoint.cartesian.y} ` : 'undefined';
```
Should there be a space in ".y} `"?

huss commented on src/client/app/utils/calibration.ts:
```diff
@@ -0,0 +1,218 @@
+
+export function isValidGPSInput(input: string) {
+	if (input.indexOf(',') === -1) { return false; }
+	const array = input.split(',').map((value: string) => parseFloat(value));
```
What happens of value is not a float so parseFloat fails? Should there be a test to verify there are only two values (meaning 1 comma) in input?

ak476519 replied:
I've added tests before these lines to make sure there are only two values(1 comma) in input.
As for this part, parseFloat will return NaN if the parsed value is not number, NaN will turn out false when compared with everything, so an invalid value will trigger the warning as expected.

huss replied:
Addressed by @ak476519. I added comment about why it works.

huss commented on src/client/app/utils/calibration.ts:
```diff
@@ -0,0 +1,218 @@
+	const deltaYInUnits = p1.cartesian.y - p2.cartesian.y;
+	const degreePerUnitY = deltaLatitude / deltaYInUnits;
+
+	const deltaLongitude = p1.gps.longitude - p2.gps.longitude;
```
Most places in the code do X first and X is the first value returned so I would put the X calculation first.

ak476519 replied:
That's fair, I put X calculation first because latitude comes before longitude.

huss replied:
Addressed by @ak476519.

huss commented on src/client/app/utils/calibration.ts:
```diff
@@ -0,0 +1,218 @@
+	if (dimensions.width > dimensions.height) {
+		const width = 500;
+		const height = 500 * dimensions.height / dimensions.width;
+		res = {
```
Could res = ... be after the if since both do this?

huss commented on src/server/models/patch-point-type.js:
```diff
@@ -0,0 +1,30 @@
+		 * @param val a string in the form "(floatA,floatB)"
+		 * @return {Point}
+		 */
+		function (val) {
```
Should this and comment above be one less indent? Seems function body should all be one more indent than function declaraton.

huss replied:
I auto-formatted to fix.

huss commented on src/server/routes/meters.js:
```diff
@@ -99,32 +101,45 @@ router.use(requiredAuthenticator);
-			.json({ message: 'Invalid meter data supplied.', err: validatorResult.errors });
+		log.warn(`Got request to edit meters with invalid meter data, errors:${validatorResult.errors}`);
+		res.status(400);
+			// .json({ message: 'Invalid meter data supplied.', err: validatorResult.errors });
```
What is the status of this comment that removed code previously there?

ak476519 replied:
It was a log.json function call that didn't seem to trigger during testing, I made a change but decided to keep the original format. Now it's been removed.

huss reviewed (changes requested):
This is nice work that does what it should. Manual testing indicates it works correctly. The automated tests worked fine. I wanted to get these comments out now but I am soon going to try to merge in development because there are conflicts. I hope to resolve the conflicts and update the PR to remove them. I’ll post another comment on this.
Here are a few general comments in addition to the ones on specific parts of the code:
- It would be nice to have more comments in the code.
- The standard check (lint) of code is generating a number of issues.
- The following files have new Meter(...) without the new gps value as an argument: src/server/services/readMacacMeters.js x 1, src/server/test/db/insertMamacReadingsFromCSVTests.js x 1, src/sever/test/db/loadFromCsvStramTests.js x 1, src/server/test/db/metasysTests.js x 3.
- This is a small point but you can enable a map that is not calibrated. It may have generated a log entry with error from the query but I’m not certain. It shows grayed out on the list in the map graphic but it might confuse an admin. Should we stop or warn an admin if they try to enable a map that is not calibrated? This can also be left for a future enhancement (we need to create an issue then).

huss committed:
Merge branch 'development' of https://github.com/OpenEnergyDashboard/OED into pr/ak476519/502

To make code backward compatible, the value of gps is set to unknown in
Meter.js. This is one of the values checked in maps to see if can
display. Also, did change the create_meters_table.sql so default
is null for similar reasons.
Not part of this PR but insertObviusData.js needed displayable set
to false for obvius meter creation.

huss committed:
items not staged in last commit due to late edit

huss committed:
fix issues with merge of development

- duplicate serverRange in graph.ts
- mistake where used unknown not undefined for meter constructor
- The upgrade of pg-promise changed the names for CTF
(Custom Type Formatting) and these were fixed.
Note this also fixed the 6 failing tests in issue: test error with
pg-promise upgrade #527.
- Change order of insert_new_map.sql to match creation order.

huss commented:
I have now completed the merging of development (see commits for what it addresses). All tests now pass and note that the changes also fixed the tests that were failing after the upgrade of pg-promise. Once the relatively straightforward comments are addressed this should be ready to merge.

Note I opened issue #532 that addresses future enhancements for maps.

ak476519 committed:
fix migrations directory

ak476519 committed:
change migration code in coherence with service

ak476519 committed:
add comment for client-side logging

ak476519 committed:
fix lint errors

ak476519 committed:
add coments to map actions

ak476519 committed:
fix lint errors

ak476519 committed:
remove obsolete comment

ak476519 committed:
add error handling to chart type recognition

ak476519 committed:
Merge branch 'mapChart' of https://github.com/ak476519/OED into mapChart

merge changes from other developers

ak476519 committed:
fix inconsistency in front-end labels

ak476519 committed:
fix error handling and lint errors

ak476519 committed:
fix source upload error handling and lint errors

ak476519 committed:
remove redundent code
used for debugging during development

ak476519 committed:
calibraton display: fix duplicate code and comment

ak476519 committed:
fix lint errors

ak476519 committed:
fix calibration algorithms

ak476519 committed:
remove outdated comments

huss committed:
Merge branch 'development' of https://github.com/OpenEnergyDashboard/OED into pr/ak476519/502

Also fixed up so all new meters have timezone then gps in creation.

huss committed:
put migrations into standard directories

huss committed:
fix linting

- Many linting issues were addressed. There are some issues
with OED so some other changes may be needed related to these.
- A couple of small mistakes from the merge of development were fixed.

huss committed:
fix unknown

This is the mistake I make when I've coded too long.
unknown -> undefined

huss committed:
miss edited line during development merge

huss committed:
missed edit in merge

huss committed:
fix circle data & PR comments

- The mapping was looping over all bars. First, it should only be using
one bar so loop removed. Second, it was using the oldest bar so now
uses the most recent one. (Actually, the bar used increased with each
meter but that was fixed when loop removed.)
- Address a few issues not done in PR comment.

huss commented:
Since many PRs were merged after the original review of this PR, I merged in development and dealt with a lot of conflicts. I also addressed a few minor outstanding comments on PR as well as fixed a bug in the values used for the circle on a map. It now seems to work correctly and locally passes all checks/tests. No issues or review items are left.
